thread 'main' panicked at 'assertion failed: self.len_after == other.len', helix-core/src/transaction.rs:116:9

### Summary

While editing, Helix crashed with the error:

```
thread 'main' panicked at 'assertion failed: self.len_after == other.len', helix-core/src/transaction.rs:116:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

### Reproduction Steps

Unfortunately I don't have steps to reproduce yet.

### Helix log

The last 50 lines are:

<details><summary>~/.cache/helix/helix.log</summary>

```
2022-12-01T09:05:06.668 helix_lsp::transport [ERROR] err <- "             , floskell\n"
2022-12-01T09:05:06.668 helix_lsp::transport [ERROR] err <- "             , ghcide-extend-import-action\n"
2022-12-01T09:05:06.668 helix_lsp::transport [ERROR] err <- "             , codeRange\n"
2022-12-01T09:05:06.668 helix_lsp::transport [ERROR] err <- "             , importLens\n"
2022-12-01T09:05:06.668 helix_lsp::transport [ERROR] err <- "             , retrie\n"
2022-12-01T09:05:06.668 helix_lsp::transport [ERROR] err <- "             , ghcide-type-lenses\n"
2022-12-01T09:05:06.668 helix_lsp::transport [ERROR] err <- "             , ghcide-code-actions-imports-exports\n"
2022-12-01T09:05:06.668 helix_lsp::transport [ERROR] err <- "             , ghcide-hover-and-symbols\n"
2022-12-01T09:05:06.668 helix_lsp::transport [ERROR] err <- "             , eval\n"
2022-12-01T09:05:06.668 helix_lsp::transport [ERROR] err <- "             , gadt\n"
2022-12-01T09:05:06.668 helix_lsp::transport [ERROR] err <- "             , callHierarchy\n"
2022-12-01T09:05:06.669 helix_lsp::transport [ERROR] err <- "             , ghcide-code-actions-type-signatures\n"
2022-12-01T09:05:06.669 helix_lsp::transport [ERROR] err <- "             , ghcide-code-actions-bindings\n"
2022-12-01T09:05:06.669 helix_lsp::transport [ERROR] err <- "             , moduleName\n"
2022-12-01T09:05:06.669 helix_lsp::transport [ERROR] err <- "             , ghcide-core\n"
2022-12-01T09:05:06.669 helix_lsp::transport [ERROR] err <- "             , explicitFixity ]\n"
2022-12-01T09:05:06.669 helix_lsp::transport [ERROR] err <- "2022-12-01T09:05:06.669175Z | Info | Logging heap statistics every 60.00s\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- " 2022-12-01T09:05:06.674245Z | Info | Starting LSP server...\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "  If you are seeing this in a terminal, you probably should have run WITHOUT the --lsp option!\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "  PluginIds: [ pragmas\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , LSPRecorderCallback\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , rename\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , ghcide-completions\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , class\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , refineImports\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , changeTypeSignature\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , alternateNumberFormat\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , ghcide-code-actions-fill-holes\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , floskell\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , ghcide-extend-import-action\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , codeRange\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , importLens\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , retrie\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , ghcide-type-lenses\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , ghcide-code-actions-imports-exports\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , ghcide-hover-and-symbols\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , eval\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , gadt\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , callHierarchy\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , ghcide-code-actions-type-signatures\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , ghcide-code-actions-bindings\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , moduleName\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , ghcide-core\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "             , explicitFixity ]\n"
2022-12-01T09:05:06.674 helix_lsp::transport [ERROR] err <- "2022-12-01T09:05:06.674481Z | Info | Starting server\n"
2022-12-01T09:05:06.675 helix_lsp::transport [ERROR] err <- "2022-12-01T09:05:06.675405Z | Info | Started LSP server in 0.00s\n"
2022-12-01T09:05:06.858 helix_lsp::transport [ERROR] err <- "2022-12-01T09:05:06.857957Z | Warning | Couldn't load cradle. Cradle not found.\n"
2022-12-01T09:05:06.858 helix_lsp::transport [ERROR] err <- "2022-12-01T09:05:06.858095Z | Warning | Currently, HLS supports GHC 9.2 only partially. See [issue #2982](https://github.com/haskell/haskell-language-server/issues/2982) for more detail.\n"
2022-12-01T09:05:06.858 helix_lsp::transport [ERROR] err <- "2022-12-01T09:05:06.858396Z | Info | Registering IDE configuration: IdeConfiguration {workspaceFolders = fromList [NormalizedUri (-7090588869910987610) \"file:///home/ollie/work/circuithub\"], clientSettings = hashed Nothing}\n"
2022-12-01T09:05:06.980 helix_lsp::transport [ERROR] err <- "2022-12-01T09:05:06.980584Z | Info | Cradle path: picofactory/pnp/lib/CircuitHub/PNP/Job/Controller.hs\n"
```
</details>


### Platform

Linux

### Terminal Emulator

Kitty

### Helix Version

22.08.1-527-g5a3ff742

--- a/helix-core/src/history.rs
+++ b/helix-core/src/history.rs
@@ -131,7 +131,7 @@ impl History {
             .map(|&n| self.revisions[n].inversion.clone());
         let down_txns = down.iter().map(|&n| self.revisions[n].transaction.clone());
 
-        up_txns.chain(down_txns).reduce(|acc, tx| tx.compose(acc))
+        down_txns.chain(up_txns).reduce(|acc, tx| tx.compose(acc))
     }
 
     /// Undo the last edit.

--- a/helix-term/tests/test/splits.rs
+++ b/helix-term/tests/test/splits.rs
@@ -161,5 +161,30 @@ async fn test_changes_in_splits_apply_to_all_views() -> anyhow::Result<()> {
     ))
     .await?;
 
+    // See <https://github.com/helix-editor/helix/issues/4957>.
+    // This sequence undoes part of the history and then adds new changes, creating a
+    // new branch in the history tree. `View::sync_changes` applies transactions down
+    // and up to the lowest common ancestor in the path between old and new revision
+    // numbers. If we apply these up/down transactions in the wrong order, this case
+    // panics.
+    // The key sequence:
+    // * 3[<space>    Create three empty lines so we are at the end of the document.
+    // * <C-w>v<C-s>  Create a split and save that point at the end of the document
+    //                in the jumplist.
+    // * <C-w>w       Switch back to the first window.
+    // * uu           Undo twice (not three times which would bring us back to the
+    //                root of the tree).
+    // * 3[<space>    Create three empty lines. Now the end of the document is past
+    //                where it was on step 1.
+    // * <C-w>q       Close window 1, focusing window 2 and causing a sync. This step
+    //                panics if we don't apply in the right order.
+    // * %d           Clean up the buffer.
+    test((
+        "#[|]#",
+        "3[<space><C-w>v<C-s><C-w>wuu3[<space><C-w>q%d",
+        "#[|]#",
+    ))
+    .await?;
+
     Ok(())
 }

